src/fin/fin_interface.cpp

```diff
@@ -0,0 +1,592 @@
+                {
+                    MIOPEN_LOG_I(
+                        "PerformanceDb: alternate record loaded: " << solver.AltSolverDbId());
+                    if(solver.IsValidPerformanceConfig(ctx, problem, config))
+                    {
+                        return config.ToString();
+                    }
+                    MIOPEN_LOG_I2("PerformanceDb: Invalid alternate record: "
+                                  << solver.AltSolverDbId() << ": " << config);
+                }
+
+                MIOPEN_LOG_I2(
+                    "PerformanceDb: Failed Loading, Using Default: " << solver.SolverDbId());
+                config = solver.GetDefaultPerformanceConfig(ctx, problem);
+                return config.ToString();
+            }
+        }
+
+        bool TestPerfCfgParams(const miopen::solver::SolverBase* solver_base,
+                               const Context& ctx,
+                               const Problem& problem,
+                               const std::string& params) const override
+        {
+            if constexpr(std::is_same_v<T, NonTunableSolver>)
+            {
+                // Non-tunable solver
+                return false;
+            }
+            else
+            {
+                // Tunable solver
+                const auto& solver = GetSolver(solver_base);
+                using PerformanceConfig =
+                    decltype(solver.GetDefaultPerformanceConfig(ctx, problem));
+                PerformanceConfig config;
+
+                if(!config.Deserialize(params))
+                {
+                    MIOPEN_LOG_WE("Perf params are obsolete or corrupt: "
+                                  << params << ". Performance may degrade.");
+                    return false;
+                }
+
+                return solver.IsValidPerformanceConfig(ctx, problem, config);
+            }
+        }
+
+        const auto& GetSolver(const miopen::solver::SolverBase* solver_base) const
+        {
+            return *static_cast<const T*>(solver_base);
+        }
+    };
+
+    template <class T>
+    void SetObject()
+    {
+        // Test cast
+        const T* ptr = dynamic_cast<const T*>(sbase);
+        if(ptr == nullptr)
+            MIOPEN_THROW(miopenStatusInternalError);
+
+        static const AnySolver_impl<T> impl;
+        obj = &impl;
+    }
+
+    void SetObjectNonTunable() { SetObject<NonTunableSolver>(); }
+
+    const AnySolver_base* obj = nullptr;
+    const miopen::solver::SolverBase* const sbase;
+};
+
+template <>
+AnySolver<miopen::ExecutionContext, miopen::conv::ProblemDescription>::AnySolver(
+    const miopen::solver::SolverBase* solver_base, uint64_t id)
+    : sbase(solver_base)
+{
+    if(!sbase->IsTunable())
+    {
+        SetObjectNonTunable();
+        return;
+    }
+
+    switch(id)
+    {
+    case 1: SetObject<miopen::solver::conv::ConvAsm3x3U>(); break;
+    case 2: SetObject<miopen::solver::conv::ConvAsm1x1U>(); break;
```

Comment:
Since it's handcrafted, I think we need some sort of instruction how to keep it synchronized (Here and in the solver counterpart, probably).

Reply:
I will move the creation of the auxiliary class into the constructor of the interface solver, so the check will occur automatically. When adding a new tunable solver without making changes to the interface, an exception will be raised, and the CI will fail.